druid/src/widget/widget.rs

```diff
@@ -217,6 +240,117 @@ impl WidgetId {
     }
 }
 
+impl WidgetPath {
+    /// Create a new `WidgetPath`.
+    ///
+    /// The provided `target` will be the only node and also the target.
+    ///
+    /// To add ancestors use the [`with_parent`] builder method to create a new path.
+    ///
+    /// To change the target use the [`with_child`] builder method to create a new path.
+    ///
+    /// [`with_parent`]: #method.with_parent
+    /// [`with_child`]: #method.with_child
+    pub fn new(target: WidgetId) -> Self {
+        Self(vec![target])
+    }
+
+    /// Add a new `parent` and return the new `WidgetPath`.
+    ///
+    /// Repeated calls with the same [`WidgetId`] will still return
+    /// a new `WidgetPath` but it will be equal in value to the current one.
+    ///
+    /// [`WidgetId`]: struct.WidgetId.html
+    #[allow(unsafe_code)]
+    pub fn with_parent(&self, parent: WidgetId) -> Self {
+        if self.0.first() == Some(&parent) {
+            self.clone()
+        } else {
+            let new_len = self.0.len() + 1;
+            let mut new = Vec::with_capacity(new_len);
+            unsafe {
+                new.set_len(new_len);
+            }
+            new[0] = parent;
+            new[1..].copy_from_slice(&self.0);
+            Self(new)
```

Comment:
I got curious about the need of `unsafe` for performance.
Why not use `Vec::extend_from_slice` which *seems* to be pretty much what you have here with a `ptr::write` instead of `copy_from_slice`
Source: https://doc.rust-lang.org/stable/src/alloc/vec.rs.html#2138

Reply:
I didn't look too deeply into `extend_from_slice` because its docs state that *this function will likely be deprecated*.

The general reason I went with the unsafe code is because I wanted to avoid iterators. Iterator code can be optimized into memcpy in some cases in the release build, but won't probably happen in debug builds. There have also been cases with even the release optimizations being broken. Although in 2020 there are now a bunch of LLVM regression tests done by the Rust team, so it's a bit more reliable.

The [extend](https://doc.rust-lang.org/stable/src/alloc/vec.rs.html#2256) special path for Copy slices seems worth looking more into though.

Overall my feeling here is that safe Rust code would be superior to unsafe usage, however not at the cost of performance, because this is a very uncomplex struct and the unsafe here isn't that unsafe.